### ISSUE TITLE
Random failures on CI related to file rename

### ISSUE THREAD

**Issue by mxr576** [composer/composer#9568]
We are running several builds (15-20) at the same time on the same server in parallel as part of our CI workflow. Some builds are randomly failing with the below-described errors. All builds are running in Docker, every build item spins up an isolated environment from the same images. The source code is mounted from the host, but Composer's cache is container only, it is not shared or mounted.

Not sure what other information can I disclose which could be useful to figure out what is failing exactly and why. Does this ring a bell to anybody?

My `composer.json`:

N/A

Output of `composer diagnose`:

```
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys: 
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
Composer version: 2.0.8
PHP version: 7.3.25
PHP binary path: /usr/local/bin/php
OpenSSL version: OpenSSL 1.1.1g  21 Apr 2020
```

When I run this command: <!-- run it with `-vvv` added to it ideally to get full debug output -->

```
composer install foo/bar
```

I get the following output: <!-- FULL OUTPUT please, not just what you think is relevant -->

```
[ErrorException]                                                                                                                                                                                                              

rename(/home/user/.composer/cache/repo/https---repo.packagist.org/provider-php-http~promise.json.tmp,/home/user/.composer/cache/repo/https---repo.packagist.org/provider-php-http~promise.json): No such file or directory  
```

**Comment by Mvannes** [comment]
Experiencing the same issue in our gitlab CI. It feels related to composer 2, as the advent of this is when we started seeing these build failures. But so far we've found no real cause as to why this could happen.

**Comment by mxr576** [comment]
Another usual error that could be related to this, Composer fails to extract downloaded artifacts. I/O or Network error? :thinking: 

![image](https://user-images.githubusercontent.com/1755573/103355439-8a4d0380-4aae-11eb-9f6c-9051f52ab876.png)

**Comment by rtm-ctrlz** [comment]
> The source code is mounted from the host

Looks like all of your concurrent jobs using same (mount from host) `vendor` directory, if it is true - I think that will have race conditions almost for every step during installs/updates.

My setup is opposite to ours:
 - separated build directory (source+vendor) for every job; reason for this - every test can do whatever it needs with sources/composer
 - shared composer cache; reason - speedup install (cache is stored at nvme-ssd) + minimize network usage (speedup + decrease load on package repos)

Also I not quite understand - how you are getting `cache-rename` issue - if every job uses it's own cache directory :(

**Comment by mxr576** [comment]
> Looks like all of your concurrent jobs using same (mount from host) vendor directory,

Damn, you are right. The initial script that creates those builds that run separately from each other uses the same Composer instance. Before Composer 2, we disabled Composer cache with `COMPOSER_CACHE_DIR=/dev/null` and now I remember, because we had similar issues.

So the https://github.com/composer/composer/issues/9568#issuecomment-752629717 is unrelated or it can be mitigated and the fix in #9580 looks good to me to fix race condition in cache writes that I originally reported here.
